### PR TITLE
AirfRANS: 5L/256d + gc=0.5 — stabilize depth scaling for extended training

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

5L/256d with golden config achieved **val_primary/surface_mse = 0.005206** at epoch 56 before catastrophically diverging at epoch 72 (grad norms: 284→641→1433→3313). This is 3x faster convergence per epoch than 4L/256d (which needed ep201 for 0.003904), suggesting 5L has superior capacity utilization — but gc=1.0 is insufficient to contain the deeper model's gradient dynamics.

**Hypothesis:** Tighter gradient clipping (gc=0.5) will prevent the ep72 divergence while preserving the fast convergence. With stability restored and 180-min budget, 5L/256d should be able to run 150+ epochs and push well below the current baseline of 0.003904.

The depth hierarchy shows non-diminishing returns: 3L→4L = +22.3%, 4L→5L = +28.3%. A stable 5L model could push below 0.003.

## Instructions

No code changes needed. Run the same golden config as the previous 5L experiment but with `--grad-clip 0.5`:

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 5 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --agent armin \
  --wandb-name "armin/airfrans-5L256d-gc05-extended" \
  --wandb-group "airfrans-depth-stability"
```

**What to report:**
1. Best val_primary/surface_mse and at which epoch
2. Did the model survive past ep72? What were grad norms at ep72?
3. Trough envelope progression (every 20 epochs)
4. Whether grad clip events changed vs gc=1.0 run

## Baseline

- **Current best:** val_primary/surface_mse = **0.003904** (PR #2755, shoya)
- **Architecture:** 4L/256d + golden config (gc=1.0, WD=1e-2, T_max=5, AdamW lr=7e-4, 223 epochs)
- **W&B run:** stxm16tv
- **External target:** 0.0043 (already beaten — now pushing below)
- **Reproduce:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT armin:

## Results

### Best metric
- **val_primary/surface_mse: 0.002851** at epoch 114
- **vs PR baseline (0.003904):** -27.0%
- **vs external target (0.0043):** -33.7%
- **vs updated radford baseline (0.00277, PR #2774 4L/256d gc=0.5):** +2.9% (slightly above)

### Surface MSE breakdown at best (ep114)
| Channel | MSE |
|---------|-----|
| Ux | 6.34e-05 |
| Uy | 5.27e-06 |
| nut | 2.96e-06 |
| p | 0.01133 |

- **full_val/volume_mse:** 0.01209

### Did the model survive past ep72?

**Yes — completely stable.** At the gc=1.0 divergence point:
- **gc=0.5 at ep72:** surface_mse=0.00672, grad_norm=7.1 (stable)
- **gc=1.0 at ep72 (prior run):** grad norms 284→641→1433→3313 (catastrophic)

gc=0.5 extended stable training from ~72 → ~167 epochs (**2.3x longer**).

### Divergence

Divergence eventually occurred at ~ep170-174:
| Epoch | grad_norm | surface_mse |
|-------|-----------|-------------|
| 166 | 4.0 | 0.00525 |
| 170 | 7.8 | 0.128 |
| 173 | 61.8 | 0.00651 |
| 174 | **589.8** | 0.0517 |
| 176 | 372.9 | 0.158 |
| 180 | 1,075 | 0.770 |
| 184 | **36,476** | 0.512 |

Checkpoint-at-best is essential. Test metrics evaluated at final (diverged) epoch are meaningless.

### Grad clip events vs gc=1.0

Grad clip events declined over training as the model learned smoother representations:
- ep1-20: 357-360/360 batches clipped (99%+)
- ep50-80: 320-350/360 (89-97%)
- ep100-140: 255-301/360 (71-84%)
- ep160+: 255-331/360 (variable before divergence)

### Trough envelope progression (every ~20 epochs)

| Epoch | surface_mse | grad_norm | vs baseline |
|-------|-------------|-----------|-------------|
| 14 | 0.0766 | 23.9 | — |
| 30 | 0.0219 | 10.1 | — |
| 44 | 0.0081 | 9.7 | — |
| 73 | 0.0052 | 6.1 | -33% |
| 89 | 0.00465 | 7.2 | — |
| 105 | **0.00341** | 5.5 | -12.7% |
| 114 | **0.00285** | 3.2 | **-27.0%** |
| 129 | 0.00297 | 3.4 | -24.0% |
| 150 | 0.00323 | 3.6 | -17.4% |
| 161 | 0.00350 | 3.1 | -10.4% |

11 total sub-0.004 readings between ep105-161 (56 epochs of consistently beating baseline).

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 5 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --agent armin \
  --wandb-name "armin/airfrans-5L256d-gc05-extended" \
  --wandb-group "airfrans-depth-stability"
```

### Run details
- **W&B run:** gvm72wj7
- **Total epochs:** 184 (180-min timeout)
- **Stable epochs:** ~167 (divergence at ~ep170)
- **Peak VRAM:** ~66.7 GB (GPU 0)
- **Epoch time:** ~0.98 min/epoch (5L/256d with Fourier)

### What happened

**gc=0.5 successfully stabilizes the 5L/256d model** — the hypothesis is confirmed. The model survived well past the ep72 divergence point and trained stably for 167 epochs with grad norms declining from ~20 to ~3 over training. The best reading of 0.002851 at ep114 beats the PR baseline by 27%.

However, **5L does not clearly outperform 4L** at gc=0.5. The updated radford baseline from PR #2774 (4L/256d gc=0.5) achieved 0.00277, which is 2.9% better than our 5L best. The 5L model converges faster per-epoch (sub-0.004 by ep105 vs ep201 for 4L gc=1.0), but the extra depth adds gradient instability that eventually causes divergence ~30 epochs earlier than 4L (ep167 vs ep201).

The depth hierarchy finding holds (5L converges faster) but the returns are diminishing — the extra layer's capacity isn't translating into deeper basins, likely because the gradient instability at cosine LR peaks limits how deep the model can explore.

### Suggested follow-ups
1. **5L/256d + gc=0.3**: Even tighter clipping might extend the stable window past ep200 and let the 5L model find deeper basins
2. **5L/256d + gc=0.5 + T_max=10**: Longer cosine cycles would reduce the frequency of LR peaks that trigger instability
3. **4L/256d + gc=0.5 + extended budget (360 min)**: Since 4L gc=0.5 already beats 5L, pushing it further with more training time might be the higher-ROI path
4. **Gradient norm warmup**: Start with very tight gc (0.1) and gradually relax to 0.5 over the first 50 epochs to build stable early representations

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2816 because ADVISOR: Closing — your 5L/256d result of 0.002851 is still far from the 0.001479 baseline. Deeper architectures (5L) are clearly not the path forward on AirfRANS. The 3L/256d sweet spot remains dominant. Thank you, armin..
